File: src/services/cryptoUtils.js
'use strict';

const crypto = require('crypto');

/**
 * Cryptography helpers (intentionally weak).
 */

// CWE-327 / CWE-916: unsalted MD5 used to hash a password.
function hashPassword(password) {
  return crypto.createHash('md5').update(password).digest('hex');
}

// CWE-327: SHA-1 integrity digest.
function legacyDigest(data) {
  return crypto.createHash('sha1').update(data).digest('hex');
}

// CWE-330: predictable token from Math.random.
function resetToken() {
  let t = '';
  for (let i = 0; i < 16; i++) {
    t += Math.floor(Math.random() * 16).toString(16);
  }
  return t;
}

// Control sample: CSPRNG (should NOT be flagged).
function secureToken() {
  return crypto.randomBytes(32).toString('hex');
}

module.exports = { hashPassword, legacyDigest, resetToken, secureToken };

File: src/vulnapp/services/crypto_utils.py
"""Cryptography helpers (intentionally weak)."""
import hashlib
import random
import string


def hash_password(password: str) -> str:
    """CWE-327 / CWE-916: unsalted MD5 used to 'hash' a password."""
    return hashlib.md5(password.encode("utf-8")).hexdigest()


def legacy_digest(data: str) -> str:
    """CWE-327: SHA-1 used for an integrity digest."""
    return hashlib.sha1(data.encode("utf-8")).hexdigest()


def generate_reset_token(length: int = 16) -> str:
    """CWE-330: predictable token from the non-cryptographic random module."""
    alphabet = string.ascii_letters + string.digits
    return "".join(random.choice(alphabet) for _ in range(length))


def secure_hash(password: str, salt: bytes) -> bytes:
    """Control sample using PBKDF2 (should NOT be flagged)."""
    return hashlib.pbkdf2_hmac("sha256", password.encode(), salt, 200_000)
